pkg/remote/gitlab/gitlab.go

```diff
@@ -0,0 +1,295 @@
+	return nil, nil
+}
+
+// Activate activates a repository by adding a Post-commit hook and
+// a Public Deploy key, if applicable.
+func (r *Gitlab) Activate(user *common.User, repo *common.Repo, keys *common.Keypair, link string) error {
+	var client = NewClient(r.URL, user.Token, r.SkipVerify)
+	var path = ns(repo.Owner, repo.Name)
+	var title, err = GetKeyTitle(link)
+	if err != nil {
+		return err
+	}
+
+	// if the repository is private we'll need
+	// to upload a github key to the repository
+	if repo.Private {
+		var err = client.AddProjectDeployKey(path, title, repo.Keys.Public)
+		if err != nil {
+			return err
+		}
+	}
+
+	// append the repo owner / name to the hook url since gitlab
+	// doesn't send this detail in the post-commit hook
+	link += "?owner=" + repo.Owner + "&name=" + repo.Name
+
+	// add the hook
+	return client.AddProjectHook(path, link, true, false, true)
+}
+
+// Deactivate removes a repository by removing all the post-commit hooks
+// which are equal to link and removing the SSH deploy key.
+func (r *Gitlab) Deactivate(user *common.User, repo *common.Repo, link string) error {
+	var client = NewClient(r.URL, user.Token, r.SkipVerify)
+	var path = ns(repo.Owner, repo.Name)
+
+	keys, err := client.ProjectDeployKeys(path)
+	if err != nil {
+		return err
+	}
+	var pubkey = strings.TrimSpace(repo.Keys.Public)
+	for _, k := range keys {
+		if pubkey == strings.TrimSpace(k.Key) {
+			if err := client.RemoveProjectDeployKey(path, strconv.Itoa(k.Id)); err != nil {
+				return err
+			}
+			break
+		}
+	}
+	hooks, err := client.ProjectHooks(path)
+	if err != nil {
+		return err
+	}
+	link += "?owner=" + repo.Owner + "&name=" + repo.Name
+	for _, h := range hooks {
+		if link == h.Url {
+			if err := client.RemoveProjectHook(path, strconv.Itoa(h.Id)); err != nil {
+				return err
+			}
+			break
+		}
+	}
+	return nil
+}
+
+// ParseHook parses the post-commit hook from the Request body
+// and returns the required data in a standard format.
+func (r *Gitlab) Hook(req *http.Request) (*common.Hook, error) {
+
+	var payload, _ = ioutil.ReadAll(req.Body)
+	var parsed, _ = gogitlab.ParseHook(payload)
+	obj := parsed.ObjectAttributes
+
+	if !(obj.State == "opened" && obj.MergeStatus == "unchecked") {
+		return nil, nil
+	}
+
+	var hook = new(common.Hook)
+
+	hook.Repo.Name = obj.Source.Name
+	hook.Repo.Owner = obj.Source.Namespace
+
+	// Check pull request comes from public fork
+	if obj.Source.VisibilityLevel < 20 {
+		//hook.SourceRemote = obj.Source.SshUrl
+		// If pull request source repo is not a public
+		// check for non-internal pull request
+		if obj.Source.Name != obj.Target.Name || obj.Source.Namespace != obj.Target.Namespace {
+			return nil, nil
+		}
+	} else {
+		//hook.SourceRemote = obj.Source.HttpUrl
+	}
+
+	hook.Commit.Author.Login = req.FormValue("owner")
+	//hook.Repo = req.FormValue("name")
+	hook.Commit.Sha = obj.LastCommit.Id
+	hook.Commit.Branch = obj.TargetBranch
+	//hook.Commit.SourceBranch = obj.SourceBranch
```

Comment:
It seems that the new data model doesn't have SourceBranch...
What should I use?